### DISCUSSION TITLE
reading from arrow dataset seems slow

### DISCUSSION THREAD

**Discussion by djouallah** [duckdb/duckdb#4509] [Q&A] [answered]
first I am using this code as a cloud function and it works, which is a miracle :)

```
import pyarrow.dataset as ds
import duckdb
import json
lineitem = ds.dataset("gs://duckddelta/delta2",format="parquet", partitioning="hive")
con = duckdb.connect()
def Query(request):
    SQL = request.get_json().get('name')
    df = con.execute(SQL).df()
    return json.dumps(df.to_json(orient="records")), 200, {'Content-Type': 'application/json'}
```

running the Query from laptop using data saved in local disk is extremely fast, so I know it is not DuckDB
based on some twitter advice !!!, it seems DuckDB is not multi threaded when getting data from an arrow dataset,  does this make any sense ?

**Comment by pdet** [comment, marked as answer]
Hi @djouallah, DuckDB should be spawning multiple threads to read arrow objects.

It does take into consideration the number of rows to know how many threads it should utilize. By default, it takes 1000000 rows per thread. i.e., if your object has 2000000, max 2 threads will start for that scan.

Having a quick look at the code, it does seem we are not filling in the object variable correctly, so this might be a bug. I'll investigate it a bit further later today.

**Comment by djouallah** [comment]
@pdet thanks :) because I got your attention, I wrote this blog https://datamonkeysite.com/2022/08/27/running-a-serverless-duckdb-on-google-cloud/ and based on the feedback people seems interested in this scenario , unfortunately to be useful DuckDB need to do some form of caching, do you think it will be possible for DuckDB to cache the data already read the first time to be used again for another Queries ?

**Reply by pdet** [reply]
I think that a way of caching this would be to create a temporary table as a select from the object, no? The only disadvantages I can think of doing so is that it becomes a pipeline breaker and (I think) that table construction is not yet parallel.

**Comment by djouallah** [comment]
@pdet ideally arrow will cache any files that was already queries, but then you need a logic to invalidate the cache if a new files showed up in the object store,  damn cache is hard, but I hope duckdb do something about it :)

**Comment by Alex-Monahan** [comment]
The ideal (fantasy?) world solution that came to my mind was some combination of checking file modified dates and tracking which rowgroups had already been loaded into RAM. Then all the pieces of the files you have already needed to query could be reused, possibly in some other combination/subset?

**Reply by pdet** [reply]
Ah, so we are talking of caching parquet files specifically?
IMO, if we do any caching of whatever underlying objects, we should already cache the duckdb memory rep of them, no?
I guess The challenge here would be keeping them up to date, but we could initially just invalidate the whole thing if modifications happened.

**Reply by Alex-Monahan** [reply]
Yes, caching the DuckDB representation would be better! But yes, throwing it all out if there was a modification handles a lot of write ~once, read many use cases. The most painful case of this is fetching data over the httpfs, which is usually parquet, but a more generic solution would be great. Is it possible DuckDB WASM has done anything in this space?

**Reply by pdet** [reply]
I'm not involved with WASM, maybe @ankoh knows?

**Reply by ankoh** [reply]
DuckDB-Wasm differs from DuckDB in that we read files in pages and cache those pages across queries.
I.e. scanning a parquet file twice over http will fetch nothing the second time.
But this is still a bit specific to the browser since we A) live shorter and B) just assume the user knows when files are no longer valid.
This assumption might be a bit dangerous in the general case, albeit maybe worth a discussion.

**Reply by djouallah** [reply]
@ankoh the case of duckdb running on a cloud function is exactly the same case, it is very short lived and downloading data from http is a pain !!!

**Comment by djouallah** [comment]
I am afraid, arrow is still substantially slower,  in this example X2

![image](https://user-images.githubusercontent.com/12554469/188024440-757db073-62bd-46c4-9516-ca277d2432a9.png)

**Reply by pdet** [reply]
Yes, we do run regression tests.

https://github.com/duckdb/duckdb/runs/8128104721?check_suite_focus=true

How many batches and what is the batch size of your lineitem object?

**Comment by djouallah** [comment]
@pdet apology, no offence intended , I don't know how to calculate the batch size ?

**Reply by pdet** [reply]
I was not offended, I was just pointing the regression example out, sorry if it sounded otherwise, text lacks nuances 😅

I think checking the batch size depends on what lineitem is, if it's a table I think there is a parameter in there, but for a stream, you might have to check each batch.

Also how many threads are you running duckdb with?

The reason I ask this is that the number of batches/threads will give us an idea on how parallel it is,  but the batches should still be roughly 10ˆ6 ~ 10^7. Otherwise, Arrow itself will present performance issues (as discussed in #3138)

**Reply by djouallah** [reply]
for the thread, I am running it by default, which should uses everything, the lineitem is a dataset, I could not find a way to get the batch size, unless I use scanner and define batch size which make the thing slower.

**Comment by djouallah** [comment]
sorry for the noise, I was confusing different issues.
using lineitem with 180 million rows in my local desktop, actually Arrow dataset is faster than Native DuckDB

```
import duckdb 
import pyarrow.dataset as ds
Lineitem = ds.dataset("/lineitem_180",format="parquet", partitioning="hive")
```

when I run using native DuckDB
```
con = duckdb.connect()
df =con.execute('''
SELECT
    --Query01
    l_returnflag,
    l_linestatus,
    SUM(l_quantity) AS sum_qty,
    SUM(l_extendedprice) AS sum_base_price,
    SUM(l_extendedprice * (1 - l_discount)) AS sum_disc_price,
    SUM(l_extendedprice * (1 - l_discount) * (1 + l_tax)) AS sum_charge,
    AVG(l_quantity) AS avg_qty,
    AVG(l_extendedprice) AS avg_price,
    AVG(l_discount) AS avg_disc,
    COUNT(*) AS count_order
FROM   'C:/lineitem_180/*/*.parquet'
WHERE
    l_shipdate <= '1998-09-02'
GROUP BY
    l_returnflag,
    l_linestatus
ORDER BY
    l_returnflag,
    l_linestatus;
''').df()
df
```
I get 40 second
using the arrow dataset, I get 32 second.

**Comment by djouallah** [comment]
@pdet I think I may have find the issue and it has nothing to do with DuckDB, it seems dataset reads the files in serial and there is no parallelism :( which make object store read very slow
https://issues.apache.org/jira/browse/ARROW-17481

**Comment by mauropagano** [comment]
@pdet would you expect that querying straight from duckdb with parquet reader is significantly slower than duckdb reading a pyarrow dataset?

**Reply by pdet** [reply]
That depends, Do you mean from an arrow dataset also reading parquet files?

If we compare an in-memory arrow dataset, it should be faster than parquet files. Since both DuckDB and Arrow have a similar memory layout, for most data types, there is no copying involved, in practice, we basically set up some pointers + makeup.

**Reply by mauropagano** [reply]
Yeah, pyarrow reading same files as pq and duckdb querying the arrow object

**Reply by pdet** [reply]
Hmm, I can't think of an immediate reason without checking the code. @hannes is our parquet reader/writer master. Maybe he has some insight on why that would be the case.
